automerge/src/types/opids.rs

```diff
@@ -0,0 +1,126 @@
+use itertools::Itertools;
+
+use super::OpId;
+
+/// A wrapper around `Vec<Opid>` which preserves the invariant that the ops are
+/// in ascending order with respect to their counters and actor IDs. In order to
+/// maintain this invariant you must provide a comparator function when adding
+/// ops as the actor indices in an  OpId are not sufficient to order the OpIds
+#[derive(Debug, Clone, PartialEq, Default)]
+pub(crate) struct OpIds(Vec<OpId>);
+
+impl<'a> IntoIterator for &'a OpIds {
+    type Item = &'a OpId;
+    type IntoIter = std::slice::Iter<'a, OpId>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.iter()
+    }
+}
+
+impl OpIds {
+    pub(crate) fn new<I: Iterator<Item = OpId>, F: Fn(&OpId, &OpId) -> std::cmp::Ordering>(
+        opids: I,
+        cmp: F,
+    ) -> Self {
+        let mut inner = opids.collect::<Vec<_>>();
+        inner.sort_by(cmp);
+        Self(inner)
+    }
+
+    /// Add an op to this set of OpIds. The `comparartor` must provide a
+    /// consistent ordering between successive calls to `add`.
+    pub(crate) fn add<F: Fn(&OpId, &OpId) -> std::cmp::Ordering>(
+        &mut self,
+        opid: OpId,
+        comparator: F,
+    ) {
+        use std::cmp::Ordering::*;
+        if self.is_empty() {
+            self.0.push(opid);
+            return;
+        }
+        let idx_and_elem = self
+            .0
+            .iter()
+            .find_position(|an_opid| matches!(comparator(an_opid, &opid), Greater | Equal));
```

Comment:
What do you think about using [binary_search_by](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.binary_search_by) here to speed it up to logn?